zkp/circuits/basetokens/anon_enc_base.circom

```diff
@@ -0,0 +1,116 @@
+// - check the input and output commitments match the expected hashes
+// - check the input and output values sum to the same amount
+// - perform encryption of the receiver's output UTXO value and salt
+template Zeto(nInputs, nOutputs) {
+  signal input inputCommitments[nInputs];
+  signal input inputValues[nInputs];
+  signal input inputSalts[nInputs];
+  // must be properly hashed and trimmed to be compatible with the BabyJub curve.
+  // Reference: https://github.com/iden3/circomlib/blob/master/test/babyjub.js#L103
+  signal input inputOwnerPrivateKey;
+  // an ephemeral private key that is used to generated the shared ECDH key for encryption
+  signal input ecdhPrivateKey;
+  signal input outputCommitments[nOutputs];
+  signal input outputValues[nOutputs];
+  signal input outputSalts[nOutputs];
+  signal input outputOwnerPublicKeys[nOutputs][2];
+  signal input encryptionNonce;
+
+  // the output for encrypted output values and salts
+  var cLen = 2 * nOutputs;
+  if (cLen % 3 != 0) {
+    cLen += (3 - (cLen % 3));
+  }
+  cLen++;
+  signal output cipherText[cLen];
+
+
+  // the output for the public key of the ephemeral private key used in generating ECDH shared key
+  signal output ecdhPublicKey[2];
+
+  // derive the sender's public key from the secret input
+  // for the sender's private key. This step demonstrates
+  // the sender really owns the private key for the input
+  // UTXOs
+  var inputOwnerPublicKey[2];
+  component pub = BabyPbk();
+  pub.in <== inputOwnerPrivateKey;
+  inputOwnerPublicKey[0] = pub.Ax;
+  inputOwnerPublicKey[1] = pub.Ay;
+  var inputOwnerPublicKeys[nInputs][2];
+  for (var i = 0; i < nInputs; i++) {
+    inputOwnerPublicKeys[i][0] = inputOwnerPublicKey[0];
+    inputOwnerPublicKeys[i][1] = inputOwnerPublicKey[1];
+  }
+
+  component checkPositives = CheckPositive(nOutputs);
+  checkPositives.outputValues <== outputValues;
+
+  component checkInputHashes = CheckHashes(nInputs);
+  checkInputHashes.commitments <== inputCommitments;
+  checkInputHashes.values <== inputValues;
+  checkInputHashes.salts <== inputSalts;
+  checkInputHashes.ownerPublicKeys <== inputOwnerPublicKeys;
+
+  component checkOutputHashes = CheckHashes(nOutputs);
+  checkOutputHashes.commitments <== outputCommitments;
+  checkOutputHashes.values <== outputValues;
+  checkOutputHashes.salts <== outputSalts;
+  checkOutputHashes.ownerPublicKeys <== outputOwnerPublicKeys;
+
+  component checkSum = CheckSum(nInputs, nOutputs);
+  checkSum.inputValues <== inputValues;
+  checkSum.outputValues <== outputValues;
+
+  // generate shared secret
+  var sharedSecret[2];
+  component ecdh = Ecdh();
+  ecdh.privKey <== ecdhPrivateKey;
+  ecdh.pubKey[0] <== outputOwnerPublicKeys[0][0];
+  ecdh.pubKey[1] <== outputOwnerPublicKeys[0][1];
+  sharedSecret[0] = ecdh.sharedKey[0];
+  sharedSecret[1] = ecdh.sharedKey[1];
+
+  // encrypt the value for the output utxos
+  component encrypt = SymmetricEncrypt(2 * nOutputs);
```

Comment:
the shared secret used to encrypt the values must be derived from the corresponding output owner's public key. as things stand now, the first receiver of the outputs (see line 95, 96) will be able to decrypt the values of all the outputs.